pro_tes/ga4gh/tes/task_runs.py

```diff
@@ -147,10 +147,12 @@ def create_task(  # pylint: disable=too-many-statements,too-many-branches
             if not is_funnel:
                 if payload_marshalled.inputs is not None:
                     for input in payload_marshalled.inputs:
-                        input.url = remove_auth_from_url(input.url)
+                        if input.url is not None:
+                            input.url = remove_auth_from_url(input.url)
                 if payload_marshalled.outputs is not None:
                     for output in payload_marshalled.outputs:
-                        output.url = remove_auth_from_url(output.url)
+                        if output.url is not None:
+                            output.url = remove_auth_from_url(output.url)
```

Comment:
So do I get this right in that Snakemake can send tasks that end up being something like `[None]`, `[None, None]`, or `["https://some.url", None]` for `payload_marshalled.inputs` and `payload_marshalled.outputs`?

If so, perhaps it's better to add something like this above line 139:

```python
# fix for clients that may generate `None`-type inputs/outputs
payload_marshalled.inputs = strip_none_items(payload_marshalled.inputs)
payload_marshalled.outputs = strip_none_items(payload_marshalled.outputs)
```

with

```python
from typing import Sequence  # put this at the top of the module, of course

def strip_none_items(_seq: Sequence) -> List:
    return list(item for item in _seq if item is not None)
```

It would be cleaner, more modular, easier to test, and probably more efficient. And by separating it from the Funnel/TESK exception code block, we wouldn't further blow up that (which we can hopefully remove at some point).

Reply:
Well, the Snakemake request loooks like this:

`{'path': '/tmp/Snakefile_s3', 'url': None, 'content': 'from snakemake.remote.S3 import RemoteProvider as S3RemoteProvider\n\nlocalrules: all, download\n\nS3 = S3RemoteProvider(access_key_id="GyCCggXpRpKQ3hBB", \n                        secret_access_key="basTuIRppYhACCdXS6yYZb1XhUTksJPq", \n                        endpoint_url="https://s3-elixir.cloud.e-infra.cz")\n\n\noutfile_s3 = "ahm2022/twardzso/outfile_smk_demo.txt"\nremote_file_s3 = "ahm2022/twardzso/news_archimedes_el.png"\n\n\nrule all:\n    input:\n        file="outfile_local.txt"\n\n\nrule getFile:\n    output:\n        file=S3.remote(remote_file_s3)\n    shell:\n        """\n        curl \'https://www.sowi.hu-berlin.de/de/lehrbereiche/s_professuren/copy_of_wissenschaftsforschung/bilder/bih-logo.png\' --output {output.file}\n        """\n\n\nrule processing:\n    input:\n        file=S3.remote(remote_file_s3)\n    output:\n        file=S3.remote(outfile_s3)\n    shell:\n        """\n        openssl dgst -md5 {input.file} > md5.txt\n        openssl dgst -sha1 {input.file} > sha1.txt\n        openssl dgst -sha224 {input.file} > sha224.txt\n        cat md5.txt sha1.txt sha224.txt > {output.file}\n        """\n\n\nrule download:\n    input:\n        file=S3.remote(outfile_s3)\n    output:\n        file="outfile_local.txt"\n    shell:\n        """\n        sleep 3\n        cp {input.file} {output.file}\n        """'}`

here  `'url': None `. In case of an execution of the task on a TESK instance   `create_task -> remove_auth_from_url` expects an URL which is not None. 
I would keep the if-clause but put it in a method and test it.

Another question. Why would you need the URL  here? 
proTES sends the task to a random TES instance which is already defined? and why would you remove the basic authentication information without saving them?
What were the usecases?

Reply:
But the list comprehension keeps the `if` clause (`if item is not None`), it just avoids an explicit loop, is more performant and less verbose. Why do you want to keep the explicit loop?

The use case is that, currently, TESK will break when you send it URLs that contain the basic auth. It is of course a problem with TESK, and should be fixed there. But it was easier to patch here. And the patch is to strip it off inputs and outputs IF the task is being sent to a TESK rather than a Funnel instance.`

Reply:
I think there is a misscommunication how the payload looks like. 
`payload` task has an attribute `input `which has the form `[item1, item2`] .

**payload_marshalled:** `Task(id=None, state=None, name='snakejob.getFile.3.sh', description='', inputs=[Input(url=None, path='/tmp/Snakefile_s3', type='FILE', name=None, description=None, content='from snakemake.remote.S3 import RemoteProvider as S3RemoteProvider\n\nlocalrules: all, download\n\nS3 = S3RemoteProvider(access_key_id="GyCCggXpRpKQ3hBB", \n                        secret_access_key="basTuIRppYhACCdXS6yYZb1XhUTksJPq", \n                        endpoint_url="https://s3-elixir.cloud.e-infra.cz")\n\n\noutfile_s3 = "ahm2022/twardzso/outfile_smk_demo.txt"\nremote_file_s3 = "ahm2022/twardzso/news_archimedes_el.png"\n\n\nrule all:\n    input:\n        file="outfile_local.txt"\n\n\nrule getFile:\n    output:\n        file=S3.remote(remote_file_s3)\n    shell:\n        """\n        curl \'https://www.sowi.hu-berlin.de/de/lehrbereiche/s_professuren/copy_of_wissenschaftsforschung/bilder/bih-logo.png\' --output {output.file}\n        """\n\n\nrule processing:\n    input:\n        file=S3.remote(remote_file_s3)\n    output:\n        file=S3.remote(outfile_s3)\n    shell:\n        """\n        openssl dgst -md5 {input.file} > md5.txt\n        openssl dgst -sha1 {input.file} > sha1.txt\n        openssl dgst -sha224 {input.file} > sha224.txt\n        cat md5.txt sha1.txt sha224.txt > {output.file}\n        """\n\n\nrule download:\n    input:\n        file=S3.remote(outfile_s3)\n    output:\n        file="outfile_local.txt"\n    shell:\n        """\n        sleep 3\n        cp {input.file} {output.file}\n        """'), Input(url=None, path='/tmp/run_snakemake.sh', type='FILE', name=None, description=None, content='#!/bin/sh\n# properties = {"type": "single", "rule": "getFile", "local": false, "input": [], "output": ["ahm2022/twardzso/news_archimedes_el.png"], "wildcards": {}, "params": {}, "log": [], "threads": 1, "resources": {"mem_mb": 1000, "mem_mib": 954, "disk_mb": 1000, "disk_mib": 954, "tmpdir": "<TBD>"}, "jobid": 3, "cluster": {}}\nmkdir /tmp/conda && cd /tmp && HOME=\'/home/valentin\' python -m snakemake --snakefile \'Snakefile_s3\' --target-jobs \'getFile:\' --allowed-rules \'getFile\' --cores \'all\' --attempt 1 --force-use-threads  --resources \'mem_mb=1000\' \'mem_mib=954\' \'disk_mb=1000\' \'disk_mib=954\'  --force --keep-target-files --keep-remote --max-inventory-time 0 --nocolor --notemp --no-hooks --nolock --ignore-incomplete --rerun-triggers \'software-env\' \'params\' \'input\' \'mtime\' \'code\' --skip-script-cleanup  --conda-frontend \'mamba\' --wrapper-prefix \'https://github.com/snakemake/snakemake-wrappers/raw/\' --latency-wait 5 --scheduler \'ilp\' --default-resources \'mem_mb=max(2*input.size_mb, 1000)\' \'disk_mb=max(2*input.size_mb, 1000)\' \'tmpdir=system_tmpdir\' --mode 2\n\n')], outputs=[], resources=Resources(cpu_cores=None, ram_gb=None, disk_gb=None, preemptible=None, zones=None), executors=[Executor(image='snakemake/snakemake:v7.24.1', command=['/bin/bash', '/tmp/run_snakemake.sh'], workdir='/tmp', stdin=None, stdout=None, stderr=None, env=None)], volumes=None, tags=None, logs=None, creation_time=datetime.datetime(2023, 3, 23, 10, 36, 48))`

The items are always non empty but the content has an URL attribute which is empty. 
That is why I added a simple if-clause to check if the URL is not None. 

Still not getting why TESK uses the URL attribute and why. It seems to behave properly when URL is absent.

Did a very small refactoring on that part of the code but left it mainly as it is.